### PR TITLE
Resolve plugins from included builds in the plugins {} block 

### PR DESCRIPTION
**Pull request by adammurdoch**
### Context

This PR adds support to the `plugins { }` build script block for resolving plugins produced by included builds within a composite build, without requiring a magic plugin resolution strategy to be defined in the consuming build.

Only those plugins declared using the `java-gradle-plugin` plugin are visible, and plugins are only visible to other builds in the same composite. They are not visible to projects in the producing build.

There are no changes for source dependencies. To use a plugin produced by a source dependency a plugin strategy is still required.

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/composite-builds/src/main/java/org/gradle/composite/internal/plugins/CompositeBuildPluginResolverContributor.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+public class CompositeBuildPluginResolverContributor implements PluginResolverContributor {
+    private final BuildStateRegistry buildRegistry;
+    private final BuildState consumingBuild;
+    private static final Resolution UNKNOWN = new Resolution(null);
```
Comment (JLLeitschuh) [inline subprojects/composite-builds/src/main/java/org/gradle/composite/internal/plugins/CompositeBuildPluginResolverContributor.java]: Weird that this is mixed in with other non-static fields. Shouldn't this be moved above or below the non-static fields?

File: subprojects/composite-builds/src/main/java/org/gradle/composite/internal/plugins/CompositeBuildPluginResolverContributor.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        resolvers.add(new CompositeBuildPluginResolver());
+    }
+
+    private class CompositeBuildPluginResolver implements PluginResolver {
+        @Override
+        public void resolve(PluginRequestInternal pluginRequest, PluginResolutionResult result) throws InvalidPluginRequestException {
+            Resolution resolution = results.get(pluginRequest.getId());
+            if (resolution != null) {
+                if (resolution.pluginResolution != null) {
+                    result.found("??", resolution.pluginResolution);
```
Comment (JLLeitschuh) [inline subprojects/composite-builds/src/main/java/org/gradle/composite/internal/plugins/CompositeBuildPluginResolverContributor.java]: What's the significance of the `"??"`?

**Comment by YeomansIII** [comment]
Does this work for the following use case?

- MainGradleModule
    - GradlePluginModule (included in settings.gradle of the Main module)

Can the Plugin that is defined using `java-gradle-plugin` in GradlePluginModule be included as a plugin `plugins { id('com.gradle.plugin') }` in the MainGradleModule?

**Comment by tprochazka** [comment]
I just tried to do that and I'm not able to make it work.

It always fails with
```
* What went wrong:
Plugin [id: 'com.myplugin'] was not found in any of the following sources:

- Gradle Core Plugins (plugin is not in 'org.gradle' namespace)
- Included Builds (None of the included builds contain this plugin)
- Plugin Repositories (plugin dependency must include a version number for this source)

```

I just put `includeBuild("plugin")` to the settings.gradle.kts and plugin folder contains plugin which define `com.myplugin`

**Pull request merged**